runtime-light/coroutine/shared-task.h

```diff
@@ -16,32 +16,32 @@
 #include "runtime-common/core/utils/kphp-assert-core.h"
 #include "runtime-light/k2-platform/k2-api.h"
 
-namespace shared_task_impl_ {
+namespace kphp::coro {
 
-// TODO: doubly-linked list to perform cancel in O(1)
-struct shared_task_waiter_t final {
+namespace shared_task_impl {
+
+struct shared_task_waiter final {
   std::coroutine_handle<> m_continuation;
-  shared_task_waiter_t *m_next{};
+  shared_task_waiter *m_next{};
+  shared_task_waiter *m_prev{};
 };
 
 template<typename promise_type>
-struct promise_base_t {
-  constexpr std::suspend_always initial_suspend() const noexcept {
+struct promise_base {
+  constexpr auto initial_suspend() const noexcept -> std::suspend_always {
     return {};
   }
 
   constexpr auto final_suspend() const noexcept {
-    struct final_suspend_t {
-      constexpr final_suspend_t() noexcept = default;
-
-      constexpr bool await_ready() const noexcept {
+    struct awaiter {
+      constexpr auto await_ready() const noexcept -> bool {
         return false;
       }
 
-      constexpr std::coroutine_handle<> await_suspend(std::coroutine_handle<promise_type> coro) const noexcept {
-        promise_base_t &promise{coro.promise()};
+      auto await_suspend(std::coroutine_handle<promise_type> coro) const noexcept -> std::coroutine_handle<> {
```

Comment:
Why constexpr has been moved?

Reply:
I decided that it's better ti give less guarantees for several functions in `shared-task.h` and `task.h`. They are actually never `constexpr`, but `constexpr` adds some limitations to the ways these functions can be used.